### PR TITLE
Support loading int8 weights

### PR DESCRIPTION
**Pull request by xiangzez**
This PR adds functions to directly load int8 model weights for int8/w8a8 datatype.

- Currently supports llama2 model. 
- Compatible with AutoGPTQ models (configuration: bits=8 and group_size=-1). Convert script will be in another PR.
- Add SrcT template type to setWeights and convertWeights function. It will be easier to add other datatypes such as int4 in the future.

### CODE REVIEW

File: src/layers/decoder_layer.h [hunk 1]
```diff
@@ -55,35 +55,69 @@ class Decoder {
...
-        const float *attnOutBias = params[7];
-        const float *gamma1 = params[8];
-        const float *beta1 = params[9];
-
-        attn.setWeights(ctx, queryWeight, queryBias, keyWeight, keyBias, valueWeight, valueBias, attnOutWeight,
-                attnOutBias, gamma1, beta1, trans);
-
-        std::vector<float *> mlpParams(params.begin() + 10, params.end());
-        mlp.setWeights(ctx, mlpParams, trans);
+    void setWeights(DecoderContext *ctx, std::vector<void *> &params, bool trans = true, int type = 0) {
```
Comment (pujiang2018) [inline src/layers/decoder_layer.h:58]: I think it is better to change "int type" to "xft::DataType dt".

File: src/layers/attention.h [hunk 2]
```diff
@@ -56,8 +56,12 @@ class Attention {
     }
 
     // The inerface is for PyTorch, thus the weights are already transposed
-    void setWeights(DecoderContext *ctx, const float *queryWeight, const float *queryBias, const float *keyWeight,
-            const float *keyBias, const float *valueWeight, const float *valueBias, const float *attnOutWeight,
+    // SrcT: float or int8_t
+    template <typename SrcT>
```
Comment (changqi1) [inline src/layers/attention.h:60]: SrcT means WeiT?

**Comment by pujiang2018** [comment]
@xiangzez a few comments. and also, you may need to rebase the code.

File: src/layers/decoder_layer.h (hunk 1 above)
Comment (changqi1) [inline src/layers/decoder_layer.h:58]: The type could use xft:data_type: https://github.com/intel/xFasterTransformer/blob/main/include/dtype.h#L18

File: src/layers/mlp_chatglm2.h [hunk 3]
```diff
@@ -22,13 +22,13 @@ class ChatGLM2MLP : public LlamaMLP<WeiT> {
     ChatGLM2MLP(DecoderContext *ctx) : LlamaMLP<WeiT>(ctx) {}
 
     // The inerface is for PyTorch, thus the weights are already transposed
-    void setWeights(DecoderContext *ctx, std::vector<float *> &params, bool trans = true) {
+    void setWeights(DecoderContext *ctx, std::vector<void *> &params, bool trans = true, int type = 0) {
```
Comment (changqi1) [inline src/layers/mlp_chatglm2.h:25]: The type could use xft:data_type

File: src/layers/mlp_llama.h [hunk 4]
```diff
@@ -41,16 +41,41 @@ class LlamaMLP : public SingletonBase<LlamaMLP<WeiT>> {
     LlamaMLP(DecoderContext *ctx) {}
 
     // The inerface is for PyTorch, thus the weights are already transposed
-    void setWeights(DecoderContext *ctx, std::vector<float *> &params, bool trans = true) {
+    void setWeights(DecoderContext *ctx, std::vector<void *> &params, bool trans = true, int type = 0) {
```
Comment (changqi1) [inline src/layers/mlp_llama.h:44]: The type could use xft:data_type

File: src/layers/mlp_llama.h [hunk 5]
```diff
@@ -41,16 +41,41 @@ class LlamaMLP : public SingletonBase<LlamaMLP<WeiT>> {
...
+            const float *normW = (const float *)params[8];
+            const int8_t *downW = (const int8_t *)params[10];
+            const float *downS = (const float *)params[11];
+            const float *downZ = (const float *)params[12];
+            setWeights(ctx, gateW, gateS, gateZ, upW, upS, upZ, normW, downW, downS, downZ, trans);
+        }
+    }
+
+    // SrcT: float or int8_t
+    template <typename SrcT>
```
Comment (changqi1) [inline src/layers/mlp_llama.h:72]: SrcT means WeiT?

File: src/models/common_decoder.h [hunk 6]
```diff
@@ -464,7 +470,7 @@ class CommonDecoder : public AbstractDecoder {
         return this->context.get();
     }
 
-    void setDecoderWeights(DECODER *pdecoder, const std::string &modelPath, int layerIdx) {
+    void setDecoderWeights(DECODER *pdecoder, const std::string &modelPath, int layerIdx, bool quant) {
```
Comment (changqi1) [inline src/models/common_decoder.h:473]: For param: bool quant, how to support fp32, bf16, fp16, int8 and int4 weights?

File: src/layers/mlp_llama.h (hunk 5 above)
Comment (xiangzez) [reply src/layers/mlp_llama.h:71]: SrcT is the source datatype of weight, convert to WeiT through MMHelper::convertWeight

**Comment by xiangzez** [comment]
PR is updated.

**Review by pujiang2018 — APPROVED** [review]

**Comment by pujiang2018** [comment]
@changqi1 pls review and help to merge.

**Comment by changqi1** [comment]
@xiangzez  Do you have some plan to support different data type model?

**Comment by xiangzez** [comment]
@changqi1 PR is updated

**Comment by changqi1** [comment]
@xiangzez  Please fix this conflict and use 2 kinds of WeiT data types to present the original weight type and converted weight type.

**Pull request merged**